### PR TITLE
Make String and Char methods like downcase and upcase unicode aware

### LINKED ISSUE #1648: Make case sensitive methods Unicode aware
Currently methods like `upcase`, `downcase` and the new `compare` only respect ASCII case sensitivity. We should add normalization and comparison methods respecting [Unicode equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence) and other case sensitivity  related parts of the Unicode standard.

### LINKED ISSUE #2439: Recognize UTF-8 whitespace characters as whitespace.
Currently crystal only recognizes ASCII whitespace characters as whitespace.
characters like \u205f,\u3000,\u0085,\u00a0,\u2007 aren't recognized. 
Looking through the source for char#whitespace? it seems like only ASCII is currently supported. Is this a design decision or is this an area that needs more work?

### PR DESCRIPTION
**Pull request by asterite**
As the title says....

Previously:

```cr
"un árbol".upcase # => "UN áRBOL"
"ЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЁЯЧСМИТЬБЮ".downcase # => "ЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЁЯЧСМИТЬБЮ"
```

Now:

```cr
"un árbol".upcase # => "UN ÁRBOL"
"ЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЁЯЧСМИТЬБЮ".downcase # => "йцукенгшщзхъфывапролджэёячсмитьбю"
```

There's also support for the special case of turkic:

```cr
"İI".downcase(Unicode::CaseOptions::Turkic) # =>"iı"
"iı".upcase(Unicode::CaseOptions::Turkic) # =>"İI"
```

This is a breaking change, because I renamed and changed some methods. Specifically, `Char#uppercase?` now checks for unicode uppercase, and there's also `Char#ascii_uppercase?` that only checks that in the ASCII range. The same applies to `whitespace?` and `ascii_whitespace?`. This shouldn't affect existing code but it might make them just a tiny bit slower.

I also renamed `Char#alpha?` to `Char#letter?` because that's the name used in unicode categories (this is the real breaking change).  Also `Char#digit?` is now `Char#number?` (again, the name used in unicode)

The unicode data is put at `src/unicode/data.cr` in the form of code, and that file is generated with a script. I used a format similar to how Go stores this data, which is quite interesting because it tries to store as little data as possible, so case ranges, alternate ranges and ascending strides are stored, and then looked up with a binary search.

I also added `String#blank?` because it's pretty useful and it has been asked a few times before.

As a side note, there's still a lot to be done regarding unicode: normalization, grapheme clusters, etc. In particular, tests found [in this article](https://mortoray.com/2013/11/27/the-string-type-is-broken/) fail in most cases in Crystal... but they also fail in Ruby, and I'm sure in other languages too (it's a hard problem, or maybe just tedious/long to implement). Elixir [does it well](http://elixir-lang.org/getting-started/binaries-strings-and-char-lists.html). But Elixir doesn't have a Char type, only String. We'd probably need to add a `Grapheme` type or similar that would be a series of unicode codepoints that all form a single visible character. But... later. This PR is a good start towards that :-)

Fixes #1648 
Fixes #2439

### CODE REVIEW

**Comment by bcardiff** [comment]
This is great. Is there a list of exceptions to handle in the future, other than Turkic?

**Comment by asterite** [comment]
@bcardiff Maybe. I took this both from [Ruby](https://github.com/ruby/ruby/blob/73199e1e82b02cd0a9c487bbee533f4398755b19/string.c#L6093-L6116) and [Go](https://github.com/golang/go/blob/7dc97d9e328edc800e2ce41d5b211ef4e0ef41d0/src/unicode/casetables.go) and turkic seems to be the only exception so far. In the future if new exceptions are added we could upgrade the code, probably handle it in a different way. In Ruby it seems you can pass multiple options so that's why I used a flags enum.

**Comment by drhuffman12** [comment]
Would it be useful to add an optional param (e.g.: something like `restrict_to` or `char_set`) to `Char#uppercase?` (and similar methods), e.g.: `Char#uppercase?(restrict_to: ASCII_SET)`,`Char#uppercase?(restrict_to: :ascii)` or `Char#uppercase?(restrict_to: CHAR::SET[:ascii])`?

**Comment by bcardiff** [comment]
I like go approach of extending the data structure by overriding some cases. But definitely something that could be refactor when we want to handle more than 1-char exception.

@drhuffman12 for the ascii string is not just to use a subset of the table but also to iterate the strings in a more efficient way. Hence the api difference.

**Comment by asterite** [comment]
Oh, yes, in fact we can add ASCII to the CaseOptions enum if you only want ascii behaviour.

**Comment by asterite** [comment]
@bcardiff Yes, Go's approach is nice but it exposes some data that might be hard to digest. Also, I don't know if in the future a crazy rule appears that doesn't match with that data definition, so I'd prefer to avoid exposing that as an API. I actually considered doing that but then noticed that it's just two special cases for turkic (the `i` and `I` letters)

**Comment by asterite** [comment]
I added an option to only transform ASCII chars

File: scripts/generate_unicode_data.cr [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    end
+  end
+
+  strides
+end
+
+entries = [] of Entry
+
+url = "http://www.unicode.org/Public/9.0.0/ucd/UnicodeData.txt"
+body = HTTP::Client.get(url).body
```
Comment (RX14) [inline scripts/generate_unicode_data.cr]: Making the crystal build process depend on an internet connection feels like a bad idea. This file should be copied into git.

Comment (kostya) [inline scripts/generate_unicode_data.cr]: this is only for generate once

**Comment by sdogruyol** [comment]
@asterite yeah Turkic is a really big problem and it's very common 😄 (i'm a Turk btw)

File: src/unicode/unicode.cr [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+# There's no need to use the methods defined in this module
+# because they are exposed in `Char` in a more convenient way
+# (`Char#upcase`, `Char#downcase`, `Char#whitespace?`, etc.)
+module Unicode
+  # Options to pass to `upcase`, `downcase`, `uppercase?`
+  # and `lowecase?` to control their behaviour.
+  @[Flags]
+  enum CaseOptions
+    # Only transform ASCII characters.
+    ASCII
```
Comment (drhuffman12) [inline src/unicode/unicode.cr]: :+1:

File: scripts/generate_unicode_data.cr (hunk 1 above)
Comment (asterite) [inline scripts/generate_unicode_data.cr]: Yes, this is generated once. Unicode data doesn't change very often so it's better here to generate it before compile-time. Generating it at compile-time, maybe with a macro, would be overkill (Elixir does that, though, but Elixir compiles modules to beam so they need to do it just once)

File: src/unicode/unicode.cr [hunk 3]
```diff
@@ -0,0 +1,137 @@
+# Provides methods that answer questions about unicode characters,
+# and the `Unicode::CaseOptions` enum for special case conversions
+# like turkic.
+#
+# There's no need to use the methods defined in this module
+# because they are exposed in `Char` in a more convenient way
+# (`Char#upcase`, `Char#downcase`, `Char#whitespace?`, etc.)
+module Unicode
+  # Options to pass to `upcase`, `downcase`, `uppercase?`
+  # and `lowecase?` to control their behaviour.
```
Comment (Sija) [inline src/unicode/unicode.cr]: typo: `lowecase?` -> `lowercase?`

Comment (asterite) [inline src/unicode/unicode.cr]: Thanks!

**Pull request merged**